certbot-dns-rfc2136/certbot_dns_rfc2136/dns_rfc2136.py

```diff
@@ -169,6 +173,30 @@ def del_txt_record(self, domain_name, record_name, record_content):
             raise errors.PluginError('Received response from server: {0}'
                                      .format(dns.rcode.to_text(rcode)))
 
+    def _get_ip_for_hostname(self, hostname):
+        """
+        Get an IP address for a supplied hostnamed.
+
+        When the system has IPv6 configured, also IPv6 adresses may be returned
+        here.
+
+        :param str hostname: The hostname to be resolved.
+        :raises OSError: if an error occurs during the lookup.
+        """
+        results = socket.getaddrinfo(hostname,  # host
+                                     53,  # port
+                                     0,  # family
+                                     socket.SOCK_DGRAM,  # type
+                                     0,  # proto
+                                     socket.AI_V4MAPPED | socket.AI_ADDRCONFIG)  # flags
+
+        # results is now a none empty list of addresses associated with the
+        # host in the order of preferences. Should no address be available, an
+        # exception would have been raised by getaddrinfo.
+
+        # Just return the first address in string representation.
+        return results[0][4][0]
```

Comment:
This is not reliable. That is... socket.gethostbyname(server) != _get_ip_for_hostname(server) 

Expecting result[0][4][0] to be valid is a mistake.

Test case, 
Install VirtualBox. (It will add multiple adaptors).
In a console window run, `ipconfig -all` and observe a number of  NICs with IPv4 and IPv6
The proposed function  _get_ip_for_hostname(server)..
a) will not reliably return an address for the NIC used for outbound network comms.
b) will only return an IPv6 address when both IPv4 and IPv6 exist

Reply:
So, the current man-page says:

```
      There are several reasons why the linked list may have  more  than  one
       addrinfo structure, including: the network host is multihomed, accessi‐
       ble over multiple protocols (e.g., both AF_INET and AF_INET6);  or  the
       same  service  is available from multiple socket types (one SOCK_STREAM
       address and another SOCK_DGRAM address, for  example).   Normally,  the
       application  should  try using the addresses in the order in which they
       are returned.   The  sorting  function  used  within  getaddrinfo()  is
       defined  in  RFC 3484; the order can be tweaked for a particular system
       by editing /etc/gai.conf (available since glibc 2.5).
```

So the first element should be the best choice when you pick only a single entry.

Since it looks like we cannot get the _best_ address for the server from the list without applying a lot of additional magic, I would recommend I change the code to return all entries. The main code will then try all entries until the DNS request succeeded for one of them. As a benefit, we can also change the configuration file in a way that you can specify multiple IP addresses there and have the additional ones as a fallback.

What do you think of that?

Reply:
I think you should try the test case for yourself as speculation may hinder your success. For example, each item returned from `iponfig all` show a valid DNS too, even if virtual. You would observe this if you tested yourself and could make a more informed design decision ;)

Reply:
It should not return an address for outgoing connections but the IP address of the Nameserver where an update is needed. it that should come from DNS, or whatever the resolver used to translate the hostname to a number.
To Help search one supplies the intended portnumber (ic. 53).

The core problem is: if socket.gethostbyname() is passed an IPv6 address it does NOT copy the address to the h_addr field. like it does for IPv4 addresses, it just fails.
therefore gethostbyname() is invalidated by reality.  (in places where there is ipv4 address available as a "backup")

So if getaddrinfo() (used by _gethostbyname(server)) cannot be used.., it would be better to drop any name lookup and just validate if it is either a valid IPv4 address or IPv6 address and create the structures themselves.
And why would socket.gethostbyname(server) return then same info as getaddrinfo(...server...)   if they allow different sorting orders and getaddrinfo().. also supplies portinfo...